removed false value for price_includes_tax and replace for public pro…

mattiLeBlanc opened the pull request:
I ran into issues while calculating tax on products for multiple tax (VAT) regions.
I noticed the taxes where always calculated like the tax was NOT included in the prices, while configuring Woocommerce taxes to INCLUDE the tax while adding products.

I traced the issue down to these 2 rules and applied a proper fix (imo).

Cheers,

Mattijs

mattiLeBlanc committed:
removed false value for price_includes_tax and replace for public propery

mikejolley commented:
Backend stored prices never include taxes.

mattiLeBlanc commented:
Sorry, but that is to short a response.
I ran into issue calculating the tax amount when enabling the setting in Woocommerce "Adding products inclusive of tax". It would ALWAYS calculate it incorrectly when I programatically add an order using the API.

I added product with a price of $15 (inclusive of tax). I have multiple VAT regions (WooCommerce EU VAT Compliance plugin).
So when I create an order, and call the $order->calculate_taxes() (based on invoice country).
It gets the correct VAT percentage, however the following thing goes wrong:

- the prices of $15 is inclusive of VAT. So WC first creates the EX-vat price (and it uses the VAT percentage of the location of the online store, Netherlands in my case, 21%).
So it calcs the price of 15/1.21 = $12.40. Then I my invoice country is Ireland (23% vat) it calculates the tax based on the $12.40, which is $15.25.

I tracked down the issue and it seems that in the lines on which I proposed the merge it always considers the prices not inclusive of tax. I using the property $this->prices_include_tax fixes my issue because now it uses the value I set in the WC settings (price inclusive of tax).

I need to point out that the price of my product is the same for all countries in Europe, however the VAT percentages varies. So for me it is paramount that the calculation of the EX-vat amount is correct and it is not with the code in the current WC Core.

If you do not accept my merge, is there another way to reach the same goal. And I do not mean circumventing the WC Tax system and using post_meta to do it myself. That would be silly.

mikejolley commented:
@mattiLeBlanc yes it was a short response because thats how prices are handled; backend prices (in the order page) are **always** store exclusive of tax. Line item taxes are stored separately.

As for the tax calculation itself, if the location being calculated is out of base location, it will remove the base taxes first, then apply the location taxes. Again, this is how core handles it, so circumventing this will cause breakage in other places which rely on said logic. Your patch will break prices for users not after your desired logic.

mattiLeBlanc commented:
Yes, that is correct it. Removes the base tax.
However the base tax is incorrect in this context because for me there is no base price without tax.
That price is based on the VAT amount.

So as I explained, we want to sell a product in Europe for 15 EUR for all countries, but each country has its own VAT amount.
So the base tax of our home location is the Netherland, 21% but that doesn't mean the 15 EUR should be broken down in a base price of 15 EUR / 1.21 and then the for example Irish tax of 23% to be applied.
That will create a inclusive price of more than 15 Eur.

I am trying to prevent us having to add a different price for each country (we use WPML so we have multilingual products) so that the base is calculated correctly.
That is a lot of administration and when VAT numbers change we have to redo that.
Not an option.

What would suggest that I do, make a custom function that does the same with the change I proposed?
Or not use the WC tax system (which would suck) and store the relevant data in post_metas?

mikejolley commented:
What you're doing is not officially supported. We've been experimenting with a 'woocommerce_adjust_non_base_location_prices' filter but its not ready for production.

In the backend as mentioned before, prices are stored as line items excl. tax. So when that excl. tax value is used to re-calc tax, it will be applied on top. So if you apply 20% tax, thats added to the excl. tax price.

This only affects backend orders.

mattiLeBlanc commented:
Actually, I was testing to find a way around and I removed my change of the core and then I saw it worked because of the change I made in my own code after the 'core hack/fix' discovery.

I added my own calculate taxes function:
```
    public function calculate_tax( $order )
    {
        $order_id   = $order->id;
        $line_item  = $order->get_items();
        $item_id    = array_keys( $line_item )[ 0 ];
        $line_item  = $line_item[ $item_id ];

        $product    = get_product( $line_item[ 'product_id'] );
        $price      = $product->get_price();

        // make sure the totals are always the same as the
        //
        wc_update_order_item_meta( $item_id, '_line_subtotal', $price );
        wc_update_order_item_meta( $item_id, '_line_total', $price  );

        $order->calculate_taxes();
        $tax        = get_post_meta( $order_id, '_order_tax', true );

        wc_update_order_item_meta( $item_id, '_line_subtotal', wc_format_decimal( $price - $tax ) );
        wc_update_order_item_meta( $item_id, '_line_total', wc_format_decimal( $price - $tax ) );

        // set the totals
        //
        $order->set_total( 0, 'cart_discount' );
        $order->set_total( 0, 'cart_discount_tax' );
        $order->set_total( $price, 'total' );
    }
```
which in combination with the standard core wc_order works fine. It calculates the price back from inclusive to exclusive for the correct VAT percentage and then it calculates the taxes so that I have a nice overview in my order.

So...nothing has to be done, I already had it working without knowing :)
